File: search/static/search/js/Treffle_Wrapper.js
class Query {
    constructor(route, query, page = 1, limit = 3, ...kwargs) {
        this.filters = {};
        this.filters_not = {}
        this.orderers = {};
        this.q = query;
        this.page = page;
        this.limit = limit;
        this.kwargs = kwargs;
        this.base_url = "https://trefle.io/api/v1";
        this.route = route;
        this.url = `${this.base_url}${this.route}`;
    }
  
    filter_(kwargs) {

        const filters = {};
        for (const [key, value] of Object.entries(kwargs)) {
            filters[key] = value;
        }
        this.filters = filters;
        console.log("filters :::", this.filters);
        return this;
      }

    filter_not(kwargs) {

        const filters_not = {};
        for (const [key, value] of Object.entries(kwargs)) {
            filters_not[key] = value;
        }
        this.filters_not = filters_not;
        console.log("filters_not :::", this.filters_not);
        return this;
    }
      
    order_by(kwargs) {
        const orderers = {}
        for (let [key, value] of Object.entries(kwargs)) {
            orderers[key] = value;
        }
        this.orderers = orderers
        return this;
    }
  
    build() {
        const params = { page: this.page };
        if (this.limit) {
            params.limit = this.limit;
        }
        if (this.q !== null) {
            params.q = this.q;
        }
        if (Object.keys(this.filters).length !== 0) {
            for (let [key, value] of Object.entries(this.filters)) {
            params[`filter[${key}]`] = value;
            }
        }
        if (Object.keys(this.filters_not).length !== 0) {
            for (let [key, value] of Object.entries(this.filters_not)) {
            params[`filter_not[${key}]`] = value;
            }
        }
        if (Object.keys(this.orderers).length !== 0) {
            for (let [key, value] of Object.entries(this.orderers)) {
            params[`order[${key}]`] = value;
            }
        }
        console.log("Filters", this.filters)
        console.log("Request Build finished}", params)
        return [this.url, params];
        }
  }
  
  class TreffleAPIWrapper {
    #token 
    #base_url
    constructor(token) {
            
        this.#base_url = "https://trefle.io/api/v1";
        this.#token = token;
    }
  
    async make_request(url, params) {

        const headers = { Authorization: `Bearer ${this.#token}` };
        try {

            console.log(`headers : ${JSON.stringify(headers)}`);
            const queryParams = new URLSearchParams(params);
            const urlWithQuery = `${url}?${queryParams}`;
            const r = await fetch(urlWithQuery, { method: "GET", headers });
                        const data = await r.json();
            if (r.status === 200) {
                return data;
            } else {
                throw new Error(
                `Request failed with status code ${r.status}. Response content: ${JSON.stringify(data)}`
                );
            }
        } catch (e) {
            console.log(e);
            return {};
        }
      }
      
  }
  
  class SearchSpecies extends Query {
    constructor(query, page = 1, limit = 5, ...kwargs) {
        const route = "/species/search";
        super(route, query, page, limit, ...kwargs);
    }
  }
  
  class SearchPlants extends Query {
    constructor(query, page = 1, limit = 3, ...kwargs) {          
        const route = "/plants/search";
        super(route, query, page, limit, ...kwargs);
    }
  }
  
  class GetItem extends Query {
    constructor(query = null, page = 1, limit = 10, ...kwargs) {
        const route = "/plants";
        super(route, query, page, limit, ...kwargs);
    }
  }
  
  //const api = new TreffleAPIWrapper("fb7c8Funa_gZnYU5onH0Oj79uapv-vvUMZ9tDqU0JTo");
  //const req = new GetItem("Passion Flower").filter_({ genus: "Passiflora"}).order_by({year : "desc"}).filter_not({year : "null"}).build()
  //console.log(req)
  //api.make_request(req[0], req[1]).then(result => console.log(result));
  export { TreffleAPIWrapper , Query , GetItem , SearchPlants, SearchSpecies}

File: Flowerz/utils/treffle_wrapper.py
import requests

TREFFLE_DATA_FIELDS = ['id', 'common_name', 'scientific_name', 'year', 'bibliography', 'author', 'family_common_name',
                       'image_url', 'synonyms', 'genus', 'family']


class Query:
    def __init__(self, route, query, page=1, limit=3, **kwargs):
        self.filters = {}
        self.orders = {}
        self.q = query
        self.page = page
        self.limit = limit
        self.kwargs = kwargs
        self.base_url = "https://trefle.io/api/v1"
        self.route = route
        self.url = "{0}{1}".format(self.base_url, self.route)

    def filter_(self, **kwargs):
        for key, value in kwargs.items():
            self.filters['{}'.format(key)] = value
        return self

    def order_by(self, **kwargs):
        for key, value in kwargs.items():
            self.orders['{}'.format(key)] = value
        return self

    def _build(self):
        params = {"page": self.page}
        if self.limit:
            params["limit"] = self.limit
        if self.q is not None:
            params["q"] = self.q
        if self.filters:
            for key, value in self.filters.items():
                params[f"filter[{key}]"] = value
        if self.orders:
            for key, value in self.orders.items():
                params[f"sort[{key}]"] = value
        return self.url, params


class TreffleAPIWrapper:
    """Wraps the Treffle api , contains subclasses of query object for each query"""

    def __init__(self, token):
        self.base_url = "https://trefle.io/api/v1"
        self.__token = token

    def make_request(self, query):
        """Make a get request to the Treffle API using the provided Query object"""
        url, params = query._build()
        print(params)
        headers = {"Authorization": f"Bearer {self.__token}"}
        try:
            r = requests.get(url=url, params=params, headers=headers)
        except requests.RequestException:
            return {}
        else:
            if r.status_code == 200:
                return r.json()
            else:
                raise Exception(f"Request failed with status code {r.status_code}. Response content: {r.content}")

    class SearchSpeciesQuery(Query):
        """Retrieve data from '/species/search' route , return Query object"""

        def __init__(self, query, page=1, limit=3, **kwargs):
            route = "/species/search"
            super().__init__(route=route, query=query, page=page, limit=limit, kwargs=kwargs)

    class SearchPlantsQuery(Query):
        """Retrieve data from '/plants/search' route , return Query object"""

        def __init__(self, query, page=1, limit=3, **kwargs):
            route = "/plants/search"
            super().__init__(route=route, query=query, page=page, limit=limit, kwargs=kwargs)

    class GetItemQuery(Query):
        """Retrieve data from '/plants/' route , return Query object"""

        def __init__(self, query=None, page=1, limit=10, **kwargs):
            route = "/plants"
            super().__init__(route=route, query=query, page=page, limit=limit, kwargs=kwargs)

    @staticmethod
    def extract_data(data, to_extract):
        extracted = []
        for flower in data:
            extracted_flower = {}
            for key in to_extract:
                if key in flower:
                    extracted_flower[key] = flower[key]
            extracted.append(extracted_flower)
        return extracted


if __name__ == "__main__":
    api = TreffleAPIWrapper("")
    # req = api.SearchSpeciesQuery(query="Passion Flower",limit=3)
    req = api.GetItemQuery(query=None, limit=5)
    print(req)
    result = api.make_request(req)
    print(len(result['data']))
